logger/syslogish/server.go

```diff
@@ -0,0 +1,148 @@
+package syslogish
+
+import (
+	"errors"
+	"fmt"
+	"log"
+	"net"
+	"regexp"
+	"strings"
+
+	"github.com/deis/deis/logger/drain"
+	"github.com/deis/deis/logger/storage"
+)
+
+var appRegex *regexp.Regexp
+
+func init() {
+	appRegex = regexp.MustCompile(`^.* ([-_a-z0-9]+)\[[a-z0-9-_\.]+\].*`)
+}
+
+// Server implements a UDP-based "syslog-like" server.  Like syslog, as described by RFC 3164, it
+// expects that each packet contains a single log message and that, conversely, log messages are
+// encapsulated in their entirety by a single packet, however, no attempt is made to parse the
+// messages received or validate that they conform to the specification.
+type Server struct {
+	conn           net.PacketConn
+	listening      bool
+	queue          chan string
+	storageAdapter storage.Adapter
+	drain          drain.LogDrain
+}
+
+// NewServer returns a pointer to a new Server instance.
+func NewServer(bindHost string, bindPort int) (*Server, error) {
+	addr, err := net.ResolveUDPAddr("udp", fmt.Sprintf("%s:%d", bindHost, bindPort))
+	if err != nil {
+		return nil, err
+	}
+	c, err := net.ListenUDP("udp", addr)
+	if err != nil {
+		return nil, err
+	}
+	return &Server{conn: c, queue: make(chan string, 5)}, nil
+}
+
+// SetStorageAdapter permits a server's underlying storage.Adapter to be reconfigured (replaced)
+// at runtime.
+func (s *Server) SetStorageAdapter(storageAdapter storage.Adapter) {
+	s.storageAdapter = storageAdapter
+}
+
+// SetDrain permits a server's underlying drain.LogDrain to be reconfigured (replaced) at runtime.
+func (s *Server) SetDrain(drain drain.LogDrain) {
+	s.drain = drain
```

Comment:
You're right.  I forgot that the nil check was there.  The drain _could_ be set to `nil` between the time it's checked and seen _not_ to be `nil` and the time the call to `d.Send()` happens.  Nice catch.  (And this really _can_ happen, because `nil` is a "valid" drain-- meaning "no drain.")  I'll fix this.  Same for the call to the storage adapter.